README.md
# psr-container-doctrine-odm: Doctrine MongoDB ODM factories for PSR-11 containers

This package provides additional factories for Doctrine MongoDB ODM on top of
[`roave/psr-container-doctrine`](https://github.com/Roave/psr-container-doctrine) to be used with containers
using the PSR-11 standard.

## Installation

```bash
$ composer require eventjet/psr-container-doctrine-odm
```

## Configuration

The most basic way is just defining the factory for the `DocumentManager`. Everything else is taken from the
configuration.

```php
return [
    'dependencies' => [
        'factories' => [
           DocumentManager::class => DocumentManagerFactory::class,
        ],
    ],
];
```

### Example Configuration

A full example of the configuration can be found in the [examples](example/full-config.php) folder.

The configuration style is heavily based
on [`roave/psr-container-doctrine`](https://github.com/Roave/psr-container-doctrine/blob/3.1.x/example/full-config.php)
to retain the familiarity. Also, the configuration _should_ be compatible with the
['DoctrineMongoODMModule for Laminas](https://github.com/doctrine/DoctrineMongoODMModule).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

File automatically generated with [phly/keep-a-changelog](https://github.com/phly/keep-a-changelog)

## 0.1.0 - 2021-08-01

This first release is actually considered stable, but it uses common factories from `roave/psr-container-doctrine`
which requires `doctrine/orm`. If you use Doctrine ORM anyway or if you don't mind having an unnecessary dependency,
go ahead :)

### Added

- [#1](https://github.com/eventjet/eventjet-psr-container-doctrine-odm/pull/1) adds all factories to create a DocumentManager

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.


tests/functional/testing.config.php
<?php

declare(strict_types=1);

use Doctrine\Persistence\Mapping\Driver\MappingDriverChain;

return [
    'doctrine' => [
        'driver' => [
            'odm_default' => [
                'class' => MappingDriverChain::class,
            ],
        ],
    ],
];


tests/functional/CreateContainerTrait.php
<?php

declare(strict_types=1);

namespace Eventjet\Test\Functional\PsrContainerDoctrineOdm;

use Doctrine\ODM\MongoDB\Configuration;
use Doctrine\ODM\MongoDB\DocumentManager;
use Eventjet\PsrContainerDoctrineOdm\Service\ConfigurationFactory;
use Eventjet\PsrContainerDoctrineOdm\Service\ConnectionFactory;
use Eventjet\PsrContainerDoctrineOdm\Service\DocumentManagerFactory;
use Eventjet\Test\Functional\PsrContainerDoctrineOdm\TestDouble\DummyPersistentCollectionFactory;
use Eventjet\Test\Functional\PsrContainerDoctrineOdm\TestDouble\DummyPersistentCollectionGenerator;
use Eventjet\Test\Functional\PsrContainerDoctrineOdm\TestDouble\DummyRepositoryFactory;
use Laminas\ServiceManager\Factory\InvokableFactory;
use Laminas\ServiceManager\ServiceManager;
use MongoDB\Client;

use function array_merge_recursive;

trait CreateContainerTrait
{
    private ?ServiceManager $container = null;
    /** @var array<string, mixed> */
    private array $additionalConfig = [];
    /** @var array<string, mixed> */
    private array $additionalServiceConfig = [];

    protected function container(): ServiceManager
    {
        if ($this->container !== null) {
            return $this->container;
        }
        $config = require __DIR__ . '/testing.config.php';
        $config = array_merge_recursive($config, $this->additionalConfig);

        $serviceConfig = array_merge_recursive(
            [
                'factories' => [
                    Configuration::class => ConfigurationFactory::class,
                    Client::class => ConnectionFactory::class,
                    DocumentManager::class => DocumentManagerFactory::class,
                    DummyPersistentCollectionFactory::class => InvokableFactory::class,
                    DummyPersistentCollectionGenerator::class => InvokableFactory::class,
                    DummyRepositoryFactory::class => InvokableFactory::class,
                ],
                'services' => [
                    'config' => $config,
                ],
            ],
            $this->additionalServiceConfig
        );
        $serviceManager = new ServiceManager($serviceConfig);
        $this->container = $serviceManager;
        return $serviceManager;
    }

    /**
     * @param array<string, mixed> $config
     */
    protected function addConfig(array $config, ?string $section = null): void
    {
        $section = $section ?? $this->configSection();
        $this->additionalConfig = array_merge_recursive(
            $this->additionalConfig,
            ['doctrine' => [$section => ['odm_default' => $config]]]
        );
    }

    /**
     * @param array<string, mixed> $config
     */
    protected function replaceConfig(array $config, string $section): void
    {
        $serviceConfig = $this->container()->get('config');
        $serviceConfig = \Safe\array_replace_recursive(
            $serviceConfig,
            ['doctrine' => [$section => ['odm_default' => $config]]]
        );
        $this->container()->setAllowOverride(true);
        $this->container()->setService('config', $serviceConfig);
        $this->container()->setAllowOverride(false);
    }

    abstract protected function configSection(): string;
}


tests/functional/TestDouble/DummyType.php
<?php

declare(strict_types=1);

namespace Eventjet\Test\Functional\PsrContainerDoctrineOdm\TestDouble;

use Doctrine\ODM\MongoDB\Types\Type;

final class DummyType extends Type
{

}


src/Service/ConnectionFactory.php
<?php

/**
 * The following class was derived from roave/psr-container-doctrine and doctrine/doctrine-mongo-odm-module
 *
 * https://github.com/Roave/psr-container-doctrine/blob/master/src/ConnectionFactory.php
 * https://github.com/doctrine/DoctrineMongoODMModule/blob/master/src/DoctrineMongoODMModule/Service/ConnectionFactory.php
 *
 * Code subject to the BSD 2-Clause license (https://github.com/Roave/psr-container-doctrine/blob/master/LICENSE)
 * and MIT License (https://github.com/doctrine/DoctrineMongoODMModule/blob/master/LICENSE), respectively
 *
 * Copyright 2016-2020 Ben Scholzen
 * Copyright 2020 Roave
 * Copyright (c) 2006-2012 Doctrine Project
 */

declare(strict_types=1);

namespace Eventjet\PsrContainerDoctrineOdm\Service;

use MongoDB\Client;
use Psr\Container\ContainerInterface;

use function is_int;
use function str_replace;
use function strpos;

use const PHP_INT_MAX;

class ConnectionFactory extends AbstractOdmFactory
{
    protected function createWithConfig(ContainerInterface $container, string $configKey): Client
    {
        $options = $this->retrieveConfig($container, $configKey, 'connection');

        $connectionString = $options['connection_string'];
        $dbName = null;

        $config = $this->retrieveDependency(
            $container,
            $options['configuration'],
            'configuration',
            ConfigurationFactory::class
        );

        $dbName = $config->getDefaultDB();

        if ($connectionString === null) {
            $connectionString = 'mongodb://';

            $user = $options['user'];
            $password = $options['password'];
            $dbName = $dbName ?? $options['dbname'];

            if ($user !== null && $password !== null) {
                $connectionString .= $user . ':' . $password . '@';
            }

            $connectionString .= $options['server'] . ':' . $options['port'];

            if ($dbName !== null) {
                $connectionString .= '/' . $dbName;
            }
        } else {
            $dbName = $this->extractDatabaseFromConnectionString($connectionString);
        }

        // Set defaultDB to $dbName, if it's not defined in configuration
        if ($dbName !== null && $config->getDefaultDB() === null) {
            $config->setDefaultDB($dbName);
        }

        $driverOptions = [];
        $driverOptions['typeMap'] = ['root' => 'array', 'document' => 'array'];

        return new Client($connectionString, $options['uri_options'], $driverOptions);
    }

    /**
     * {@inheritdoc}
     */
    protected function getDefaultConfig(string $configKey): array
    {
        return [
            'server' => 'localhost',
            'port' => '27017',
            'user' => null,
            'password' => null,
            'dbname' => null,
            'connection_string' => null,
            'uri_options' => [],
            'configuration' => $configKey,
        ];
    }

    private function extractDatabaseFromConnectionString(string $connectionString): ?string
    {
        $dbName = null;
        $connectionString = str_replace('mongodb://', '', $connectionString);
        $dbStart = strpos($connectionString, '/');
        if (!is_int($dbStart)) {
            return null;
        }
        $dbEnd = strpos($connectionString, '?');
        $dbName = \Safe\substr(
            $connectionString,
            $dbStart + 1,
            $dbEnd !== false ? ($dbEnd - $dbStart - 1) : PHP_INT_MAX
        );
        return $dbName;
    }
}


src/Service/ConfigurationFactory.php
<?php

/**
 * The following class was derived from roave/psr-container-doctrine and doctrine/doctrine-mongo-odm-module
 *
 * https://github.com/Roave/psr-container-doctrine/blob/master/src/ConfigurationFactory.php
 * https://github.com/doctrine/DoctrineMongoODMModule/blob/master/src/DoctrineMongoODMModule/Service/ConfigurationFactory.php
 *
 * Code subject to the BSD 2-Clause license (https://github.com/Roave/psr-container-doctrine/blob/master/LICENSE)
 * and MIT License (https://github.com/doctrine/DoctrineMongoODMModule/blob/master/LICENSE), respectively
 *
 * Copyright 2016-2020 Ben Scholzen
 * Copyright 2020 Roave
 * Copyright (c) 2006-2012 Doctrine Project
 */

declare(strict_types=1);

namespace Eventjet\PsrContainerDoctrineOdm\Service;

use Doctrine\ODM\MongoDB\Configuration;
use Doctrine\ODM\MongoDB\Repository\DocumentRepository as DefaultDocumentRepository;
use Doctrine\ODM\MongoDB\Types\Type;
use Psr\Container\ContainerInterface;
use Roave\PsrContainerDoctrine\CacheFactory;
use Roave\PsrContainerDoctrine\DriverFactory;

class ConfigurationFactory extends AbstractOdmFactory
{
    protected function createWithConfig(ContainerInterface $container, string $configKey): Configuration
    {
        $options = $this->retrieveConfig($container, $configKey, 'configuration');

        $config = new Configuration();

        // proxies
        $config->setProxyDir($options['proxy_dir']);
        $config->setProxyNamespace($options['proxy_namespace']);
        $config->setAutoGenerateProxyClasses($options['generate_proxies']);

        // hydrators
        $config->setAutoGenerateHydratorClasses($options['generate_hydrators']);
        $config->setHydratorDir($options['hydrator_dir']);
        $config->setHydratorNamespace($options['hydrator_namespace']);

        // persistent collections
        $config->setAutoGeneratePersistentCollectionClasses($options['generate_persistent_collections']);
        $config->setPersistentCollectionDir($options['persistent_collection_dir']);
        $config->setPersistentCollectionNamespace($options['persistent_collection_namespace']);

        if ($options['persistent_collection_factory'] !== null) {
            $config->setPersistentCollectionFactory($container->get($options['persistent_collection_factory']));
        }

        if ($options['persistent_collection_generator'] !== null) {
            $config->setPersistentCollectionGenerator(
                $container->get($options['persistent_collection_generator'])
            );
        }

        // default db
        if ($options['default_db'] !== null) {
            $config->setDefaultDB($options['default_db']);
        }

        // caching
        $config->setMetadataCacheImpl(
            $this->retrieveDependency(
                $container,
                $options['metadata_cache'],
                'cache',
                CacheFactory::class
            )
        );

        // Register filters
        foreach ($options['filters'] as $alias => $class) {
            $config->addFilter($alias, $class);
        }

        // the driver
        $config->setMetadataDriverImpl(
            $this->retrieveDependency(
                $container,
                $options['driver'],
                'driver',
                DriverFactory::class
            )
        );

        // metadataFactory, if set
        if ($options['class_metadata_factory_name'] !== null) {
            $config->setClassMetadataFactoryName($options['class_metadata_factory_name']);
        }

        // respositoryFactory, if set
        if ($options['repository_factory'] !== null) {
            $config->setRepositoryFactory($container->get($options['repository_factory']));
        }

        // custom types
        foreach ($options['types'] as $name => $class) {
            if (Type::hasType($name)) {
                Type::overrideType($name, $class);
            } else {
                Type::addType($name, $class);
            }
        }

        $config->setDefaultDocumentRepositoryClassName($options['default_document_repository_class_name']);

        return $config;
    }

    /**
     * {@inheritdoc}
     */
    protected function getDefaultConfig(string $configKey): array
    {
        return [
            'metadata_cache' => 'array',
            'generate_proxies' => Configuration::AUTOGENERATE_EVAL,
            'proxy_dir' => 'data',
            'proxy_namespace' => 'DoctrineMongoODMModule\Proxy',
            'generate_hydrators' => Configuration::AUTOGENERATE_ALWAYS,
            'hydrator_dir' => 'data',
            'hydrator_namespace' => 'DoctrineMongoODMModule\Hydrator',
            'generate_persistent_collections' => Configuration::AUTOGENERATE_ALWAYS,
            'persistent_collection_dir' => 'data',
            'persistent_collection_namespace' => 'DoctrineMongoODMModule\PersistentCollection',
            'persistent_collection_factory' => null,
            'persistent_collection_generator' => null,
            'default_db' => null,
            'filters' => [],
            'class_metadata_factory_name' => null,
            'repository_factory' => null,
            'types' => [],
            'driver' => $configKey,
            'default_document_repository_class_name' => DefaultDocumentRepository::class,
        ];
    }
}


src/Service/DocumentManagerFactory.php
<?php

/**
 * The following class was derived from roave/psr-container-doctrine and doctrine/doctrine-mongo-odm-module
 *
 * https://github.com/doctrine/DoctrineMongoODMModule/blob/master/src/DoctrineMongoODMModule/Service/DocumentManagerFactory.php
 *
 * Code subject to the BSD 2-Clause license (https://github.com/Roave/psr-container-doctrine/blob/master/LICENSE)
 * and MIT License (https://github.com/doctrine/DoctrineMongoODMModule/blob/master/LICENSE), respectively
 *
 * Copyright 2016-2020 Ben Scholzen
 * Copyright 2020 Roave
 * Copyright (c) 2006-2012 Doctrine Project
 */

declare(strict_types=1);

namespace Eventjet\PsrContainerDoctrineOdm\Service;

use Doctrine\ODM\MongoDB\DocumentManager;
use Psr\Container\ContainerInterface;
use Roave\PsrContainerDoctrine\EventManagerFactory;

class DocumentManagerFactory extends AbstractOdmFactory
{
    /**
     * {@inheritdoc}
     */
    protected function createWithConfig(ContainerInterface $container, string $configKey)
    {
        $options = $this->retrieveConfig($container, $configKey, 'documentmanager');

        $connection = $this->retrieveDependency(
            $container,
            $options['connection'],
            'connection',
            ConnectionFactory::class
        );
        $configuration = $this->retrieveDependency(
            $container,
            $options['configuration'],
            'configuration',
            ConfigurationFactory::class
        );
        $eventManager = $this->retrieveDependency(
            $container,
            $options['event_manager'],
            'event_manager',
            EventManagerFactory::class
        );

        return DocumentManager::create($connection, $configuration, $eventManager);
    }

    /**
     * {@inheritdoc}
     */
    protected function getDefaultConfig(string $configKey): array
    {
        return [
            'connection' => $configKey,
            'configuration' => $configKey,
            'event_manager' => $configKey,
        ];
    }
}


example/full-config.php
<?php

/**
 * This configuration is based on the config of roave/psr-container-doctrine to retain familiarity with its style.
 * Factories which are used from roave/psr-container-doctrine
 * (namely the \Roave\PsrContainerDoctrine\EventManagerFactory) have the same configuration except that the default
 * 'odm_default' is used.
 *
 * @see https://github.com/Roave/psr-container-doctrine/blob/3.1.x/example/full-config.php
 */

declare(strict_types=1);

use Doctrine\ODM\MongoDB\Configuration;
use Doctrine\ODM\MongoDB\Repository\DocumentRepository as DefaultDocumentRepository;

return [
    'doctrine' => [
        'configuration' => [
            'odm_default' => [
                'metadata_cache' => 'array',
                'generate_proxies' => Configuration::AUTOGENERATE_EVAL,
                'proxy_dir' => 'data',
                'proxy_namespace' => 'DoctrineMongoODMModule\Proxy',
                'generate_hydrators' => Configuration::AUTOGENERATE_ALWAYS,
                'hydrator_dir' => 'data',
                'hydrator_namespace' => 'DoctrineMongoODMModule\Hydrator',
                'generate_persistent_collections' => Configuration::AUTOGENERATE_ALWAYS,
                'persistent_collection_dir' => 'data',
                'persistent_collection_namespace' => 'DoctrineMongoODMModule\PersistentCollection',
                'persistent_collection_factory' => null,
                'persistent_collection_generator' => null,
                'default_db' => null,
                'filters' => [],
                'class_metadata_factory_name' => null,
                'repository_factory' => null,
                'types' => [],
                'driver' => 'odm_default', // Actually defaults to the configuration config key, not hard-coded
                'default_document_repository_class_name' => DefaultDocumentRepository::class,
            ],
        ],
        'connection' => [
            'odm_default' => [
                'server' => 'localhost',
                'port' => '27017',
                'user' => null,
                'password' => null,
                'dbname' => null,
                'connection_string' => null,
                'uri_options' => [],
                'configuration' => 'odm_default', // Actually defaults to the configuration config key, not hard-coded
            ],
        ],
    ],
];
